main.go

```diff
@@ -0,0 +1,229 @@
+	"github.com/sirupsen/logrus"
+)
+
+var configurationPath = "/opt/kubeadm"
+
+func clusterProvider(cluster clusterplugin.Cluster) yip.YipConfig {
+
+	// TODO(kpiyush17) add cluster.Options handling, expecting the options to match KubeadmConfig spec
+
+	var stages []yip.Stage
+
+	preStage := yip.Stage{
+		Systemctl: yip.Systemctl{
+			Enable: []string{"kubelet"},
+			Start:  []string{"containerd"},
+		},
+		Commands: []string{"sudo sysctl --system"},
+	}
+
+	stages = append(stages, preStage)
+
+	if cluster.Role == clusterplugin.RoleInit {
+		stages = append(stages, getInitYipStages(cluster)...)
+	} else if (cluster.Role == clusterplugin.RoleControlPlane) || (cluster.Role == clusterplugin.RoleWorker) {
+		stages = append(stages, getJoinYipStages(cluster)...)
+	}
+
+	cfg := yip.YipConfig{
+		Name: "Kubeadm C3OS Cluster Provider",
+		Stages: map[string][]yip.Stage{
+			"boot.before": stages,
+		},
+	}
+
+	return cfg
+}
+
+func main() {
+	plugin := clusterplugin.ClusterPlugin{
+		Provider: clusterProvider,
+	}
+
+	if err := plugin.Run(); err != nil {
+		logrus.Fatal(err)
+	}
+}
+
+func getInitYipStages(cluster clusterplugin.Cluster) []yip.Stage {
+	kubeadmCfg := getInitNodeConfiguration(cluster)
+	return []yip.Stage{
+		{
+			Name: "Install Kubeadm Configuration Files",
+			Files: []yip.File{
+				{
+					Path:        filepath.Join(configurationPath, "kubeadm.yaml"),
+					Permissions: 0640,
+					Content:     kubeadmCfg,
+				},
+			},
+		},
+		{
+			Name: "Kubeadm init",
+			Commands: []string{
+				fmt.Sprintf("kubeadm init --config %s --upload-certs", filepath.Join(configurationPath, "kubeadm.yaml")),
+			},
+		},
+		{
+			Name: "Kubeadm export kubeconfig",
+			Commands: []string{
+				fmt.Sprintf("export KUBECONFIG=%s", "/etc/kubernetes/admin.conf"),
+			},
+		},
+		{
+			Name: "Post-Kubeadm apply Calico CNI",
+			Commands: []string{
+				fmt.Sprintf("kubectl apply -f %s -n kube-system", filepath.Join(configurationPath, "calico.yaml")),
+			},
+		},
+	}
+}
+
+func getJoinYipStages(cluster clusterplugin.Cluster) []yip.Stage {
+	kubeadmCfg := getJoinNodeConfiguration(cluster)
+
+	joinCmd := fmt.Sprintf("kubeadm join --config %s", filepath.Join(configurationPath, "kubeadm.yaml"))
+	if cluster.Role == clusterplugin.RoleControlPlane {
+		joinCmd = joinCmd + " --control-plane"
+	}
+
+	return []yip.Stage{
+		{
+			Name: "Install Kubeadm Configuration Files",
+			Files: []yip.File{
+				{
+					Path:        filepath.Join(configurationPath, "kubeadm.yaml"),
+					Permissions: 0640,
+					Content:     kubeadmCfg,
+				},
+			},
+		},
+		{
+			Name: "Kubeadm join control plane",
+			Commands: []string{
+				joinCmd,
+			},
+		},
+	}
+}
+
+func getInitNodeConfiguration(cluster clusterplugin.Cluster) string {
+	certificateKey := getCertificateKey(cluster.ClusterToken)
+
+	var initCfg bytes.Buffer
+	_ = yaml.NewEncoder(&initCfg).Encode(config.InitConfiguration{
+		TypeMeta: config.TypeMeta{
+			Kind:       "InitConfiguration",
```

Comment:
it would be better to use `k8s.io/cli-runtime/pkg/printers` instead of hard coding this meta